src/modal/modal.svelte

```diff
@@ -0,0 +1,44 @@
+<script>
+  import { getContext, onMount, onDestroy } from 'svelte';
+  import modalContextKey from './modal-context-key.js';
+
+  const { register, update, unregister } = getContext(modalContextKey);
+
+  export let open = false;
+
+  /* The component to use in creating a modal. */
+  export let component;
```

Comment:
I really don't like the idea of taking a component as a prop instead of a slot. I see only 2 ways around it:
using `$$props.$$slots`, which works but isn't part of the official API (although it **is** used in some libraries. Check out https://github.com/sveltejs/svelte/issues/2106#issuecomment-501000254 for an example and the rest of the discussion) but would make more sense for the user; it will `console.warn` that the component doesn't expect a slot, though. The other way would be to wrap a slot in a meaningless `div` just to have something to `bind:this` to, and use that as the component. This seems like the more resilient way, although it introduces a useless element

Reply:
Me neither, but that is the only way we can have the `<Modal>` instances anywhere in the descendants of the `<ModalOverlay>`, which was a first requirement. Otherwise it would create this slotted component right where it is placed.
Some might argue something like a _portal_ could remedy that, but no such thing exists in Svelte as of now, and doing it manually with `appendChild` is a sure way for disaster.

Not only is `$$props.$$slots` internal API, the use of `$$props` themselves is discouraged by Svelte.

As a final note, given the experience of writing a shit ton of modals in the View Project page of Innopoints, I can assure you that you always want the modal window to be a separate component. Simply because it contains some sort of logic that is best isolated away.

Reply:
But don't you think that this way is much more limiting than using `$$props.$$slots` (which was even sorta suggested by a Svelte member: https://github.com/sveltejs/rfcs/pull/15#issuecomment-570811127) or using a dummy wrapper with `bind:this`? You are only passing props (in a way that looks weird, tbh), which doesn't allow listening for events or adding directives

Reply:
> But don't you think that this way is much more limiting than using `$$props.$$slots`...

Yes, it is more limiting. But again, that's okay because you want the modal created in a different place, in fact. And because most modals require a component of their own for convenience anyway. This is a limitation that we cannot get around with, for if we go with `$$props.$$slots`, we will end up creating the component further down the element tree, and our only option would be to move it.

> You are only passing props (in a way that looks weird, tbh), which doesn't allow listening for events or adding directives

Not exactly. Remember the events action? Here's your way to listen to events. As for directives, you cannot add them to a custom component anyway (I think), and even if you'd want to, you could do that in the component itself, to one of its elements, and then pass that component here.